File: tutorial3-complete/src-ios/Libraries/FoyerKit/Sources/MainShell/MainBuilder.swift
// Copyright (c) 2026 Modaal.dev
// Licensed under the MIT License. See LICENSE file for details.

import FoyerBridge
import FoyerKit
import Foundation
import HomeShell
import ProfileShell

/// What the main level consumes: the union of what its two tabs' subtrees need.
/// sourcery: DuetComponent
/// sourcery: CreateMock
public protocol MainDependency: AnyObject {
  var items: any ItemsPort { get }
  var auth: any AuthPort { get }
  var account: any AccountPort { get }
}

extension MainComponent: HomeDependency {}
extension MainComponent: ProfileDependency {}

final class LiveMainEnvironment: NSObject, MainEnvironment {
  private let onDelegate: (MainDelegateEvent) -> Void

  init(onDelegate: @escaping (MainDelegateEvent) -> Void) {
    self.onDelegate = onDelegate
  }

  func notifyHost(event: MainDelegateEvent) {
    onDelegate(event)
  }
}

extension MainComponent {
  @MainActor
  func environment(onDelegate: @escaping (MainDelegateEvent) -> Void) -> LiveMainEnvironment {
    LiveMainEnvironment(onDelegate: onDelegate)
  }
}

// MARK: - Builder

public final class MainChild {
  public let shell: MainViewShell

  init(shell: MainViewShell) {
    self.shell = shell
  }
}

public final class MainBuilder {
  private let dependency: MainDependency

  public init(dependency: MainDependency) {
    self.dependency = dependency
  }

  /// Builds the main mount and both tabs: they live for the level's
  /// lifetime, so the Builder constructs them here over the same per-mount
  /// Component, and the shell brackets them with its own activation.
  @MainActor
  public func buildMain(
    displayName: String,
    onDelegate: @escaping (MainDelegateEvent) -> Void
  ) -> MainChild {
    let component = MainComponent(dependency: dependency)
    let scope = mainImmediateStoreScope()
    let store = makeMainStore(
      environment: component.environment(onDelegate: onDelegate),
      scope: scope)
    let bridged = MainKitStore(
      state: mainStateFlow(store: store),
      send: { store.send(action: $0) },
      teardown: {
        store.teardown()
        cancelStoreScope(scope: scope)
      })
    let shell = MainViewShell(store: bridged)
    let home = HomeBuilder(dependency: component).buildHome()
    let profile = ProfileBuilder(dependency: component)
      .buildProfile(displayName: displayName) { [weak shell] event in
        shell?.store.send(MainActionProfile(event: event))
      }
    shell.attach(home: home, profile: profile)
    return MainChild(shell: shell)
  }
}

File: tutorial4-start/src-kmp/app/src/main/kotlin/dev/modaal/foyer/app/MainBuilder.kt
// Copyright (c) 2026 Modaal.dev
// Licensed under the MIT License. See LICENSE file for details.

package dev.modaal.foyer.app

import dev.modaal.duet.kernel.Store
import dev.modaal.duet.shells.StoreHost
import dev.modaal.foyer.main.MainAction
import dev.modaal.foyer.main.MainDelegateEvent
import dev.modaal.foyer.main.MainEffectPayload
import dev.modaal.foyer.main.MainEnvironment
import dev.modaal.foyer.main.MainState
import dev.modaal.foyer.main.makeMainStore
import dev.modaal.foyer.ports.AccountPort
import dev.modaal.foyer.ports.AuthPort
import dev.modaal.foyer.ports.ItemsPort
import kotlinx.coroutines.CoroutineScope

typealias MainStore = Store<MainState, MainAction, MainEffectPayload>

/** What the main level consumes: the union of what its two tabs' subtrees need. */
interface MainDependency {
  val items: ItemsPort
  val auth: AuthPort
  val account: AccountPort
}

/** Forwards the Dependency and satisfies both tabs' Dependencies with the same members. */
class MainComponent(dependency: MainDependency) :
  MainDependency by dependency, HomeDependency, ProfileDependency {
  fun environment(onDelegate: (MainDelegateEvent) -> Unit): MainEnvironment =
    object : MainEnvironment {
      override fun notifyHost(event: MainDelegateEvent) = onDelegate(event)
    }
}

/**
 * What one main mount owns: its store and both tabs, mounted for the level's
 * lifetime. The tab bar chooses which one is shown; neither is rebuilt on a
 * switch.
 */
class MainMount(
  val store: MainStore,
  val home: HomeStore,
  val profile: ProfileMount,
  private val host: StoreHost,
) {
  fun teardown() = host.teardownAll()
}

class MainBuilder(private val dependency: MainDependency) {
  fun buildMain(
    displayName: String,
    onDelegate: (MainDelegateEvent) -> Unit,
    scope: CoroutineScope,
  ): MainMount {
    val component = MainComponent(dependency)
    val host = StoreHost(scope)
    val store = host.host(makeMainStore(environment = component.environment(onDelegate), scope = scope))
    val home = host.host(HomeBuilder(component).buildHome(scope))
    val profile =
      ProfileBuilder(component)
        .buildProfile(
          displayName = displayName,
          onDelegate = { store.send(MainAction.Profile(it)) },
          scope = scope,
        )
    host.adoptTeardown(profile::teardown)
    return MainMount(store, home, profile, host)
  }
}
